beacon-chain/rpc/validator/attester.go

```diff
@@ -84,6 +92,32 @@ func (vs *Server) GetAttestationData(ctx context.Context, req *ethpb.Attestation
 		return nil, status.Errorf(codes.Internal, "Could not retrieve head root: %v", err)
 	}
 
+	// In the case that we receive an attestation request after a newer state/block has been
+	// processed, we walk up the chain until state.Slot <= req.Slot to prevent producing an
+	// attestation that violates processing constraints.
+	for headState.Slot() > req.Slot {
+		if ctx.Err() != nil {
+			return nil, status.Errorf(codes.Aborted, ctx.Err().Error())
+		}
+		parent := headState.ParentRoot()
+		headRoot = parent[:]
+		headState, err = vs.BeaconDB.State(ctx, parent)
```

Comment:
Optional comment. I like the pattern below. So when we deprecate `NewStateMgmt`, this becomes much easier to clean up (ie just delete the else block)

```go
if featureconfig.Get().NewStateMgmt {
	headState, err = vs.StateGen.StateByRoot(ctx, parent)
	if err != nil {
		return nil, status.Errorf(codes.Internal, fmt.Sprintf("Failed to generate state: %v", err))
	}
} else {
      headState, err = vs.BeaconDB.State(ctx, parent)
       if err != nil {
			return nil, status.Error(codes.Internal, err.Error())
	}
}
if headState == nil {
	return nil, status.Error(codes.Internal, "Failed to lookup parent state from head.")
}
```

Reply:
Hmm, but what if the db.State returns nil? I would still need to call state gen?